pkg/registry/generic/registry/store.go

```diff
@@ -412,12 +522,20 @@ func (e *Store) Delete(ctx api.Context, name string, options *api.DeleteOptions)
 	if err != nil {
 		return nil, err
 	}
+	// this means finalizers cannot be updated via DeleteOptions if a deletion is already pending
 	if pendingGraceful {
 		return e.finalizeDelete(obj, false)
 	}
+	// check if obj has pending finalizers
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return nil, kubeerr.NewInternalError(err)
+	}
+	pendingFinalizers := len(accessor.GetFinalizers()) != 0
 	var ignoreNotFound bool = false
 	var lastExisting runtime.Object = nil
-	if graceful {
+	// update the object
+	if graceful || (EnableGarbageCollector && (pendingFinalizers || shouldUpdateFinalizers(accessor, options))) {
```

Comment:
I'm not confident that EnableGarbageCollector is handled right in this block of code, and this isn't very testable. Consider just copy-pasting. Even better if you can move this chunk of work into its own function? then just have two versions of the function.